apps/src/labs/types.ts

```diff
@@ -0,0 +1,130 @@
+// TODO: other channel properties mentioned in project.js:
+// level, frozen, hidden, thumbnailUrl, migratedToS3, sharedWith, libraryName, libraryDescription,
+// latestLibraryVersion, publishLibrary, libraryPublishedAt
+//
+// Do we still want/need these? Should they be on a separate type?
+// If the ChannelsApi on the server doesn't care about these, they should
+// live elsewhere.
+// The library data should definitely live elsewhere.
+export interface Channel {
+  id: string;
+  name: string;
+  isOwner: boolean;
+  projectType: ProjectType | null;
+  publishedAt: string;
+  createdAt: string;
+  updatedAt: string;
+}
+
+export type DefaultChannel = Pick<Channel, 'name'>;
+
+// We will eventually make this a union type to include other source types.
+export type Source = BlocklySource;
+
+export interface SourceUpdateOptions {
+  currentVersion: string;
+  replace: boolean;
+  firstSaveTimestamp: string;
+  tabId: string | null;
+}
+
+export interface Project {
+  source: Source;
+  channel: Channel;
+}
+
+export interface BlocklySource {
+  blocks: {
+    languageVersion: number;
+    blocks: BlocklyBlock[];
+  };
+  variables: BlocklyVariable[];
+}
+
+export interface BlocklyBlock {
+  type: string;
+  id: string;
+  x: number;
+  y: number;
+  next: {
+    block: BlocklyBlock;
+  };
+}
+
+export interface BlocklyVariable {
+  name: string;
+  id: string;
+}
+
+// TODO: these are not all the properties of the level.
+// Fill this in as we need them.
+export interface Level {
+  projectType: ProjectType;
+  isK1: boolean;
+  standaloneAppName: StandaloneAppName;
+  useContractEditor: boolean;
+  // Minecraft specific properties
+  isAgentLevel: boolean;
+  isEventLevel: boolean;
+  isConnectionLevel: boolean;
+  isAquaticLevel: boolean;
+}
+
+// TODO: these are not all the properties of app options.
+// Fill this in as we need them.
+export interface AppOptions {
+  app: AppName;
+  level: Level;
+  skinId: string;
+  droplet: boolean;
+  channel: string;
+}
+
+export type ProjectType =
```

Comment:
Is there a way to centralize these lists between client & server?  It feels like our source tree has similar lists in quite a few different places.

Reply:
I don't think we can share types between ruby and typescript. We do have shared constants that we generate, but I don't think it would give us typescript types. The hope here is we can at least have this be a single source of truth on the frontend.